PROTOSPEC.md

```diff
@@ -0,0 +1,62 @@
+# Specification outline
+
+An outline of the logic for this proposal, in the way that it would be specified:
+
+## Grammar
+
+```
+DecoratorIdentifier :: @ IdentifierName
+
+Decorator : DecoratorIdentifier ArgumentsList_opt
+
+DecoratorList : Decorator+
+```
+
+`DecoratorList`s can be used as in the current decorators proposal: decorating classes, methods, or field declarations.
+
+```
+DecoratorDeclaration :
+  decorator [no LineTerminator here] DecoratorIdentifier (FormalParameters[~Yield, ~Await])_opt { DecoratorList }
+```
+
+## Scopes
+
+Similar to `#names`, there is a parallel lexical scope for `@names`. This is important so that `with` statements do not provide a way to dynamically create decorators.
+
+Each construct which makes a new lexical scope for variables also names a lexical scope for decorators. The only type of declarations which decorators support are `const` declarations and `import { @name }` declarations.
+
+There is a special outer scope for built-in decorators. When a decorator is declared at the module or script level, it is being declared in an inner, nested scope which may shadow things in this outer scope. In this way, built-in decorators can be polyfilled.
+
+## Early errors
+
+Although the decorators specification is designed to allow decorators to be analyzed ahead of time, the JavaScript specification is organized with everything running at runtime, except for analysis of early errors. Because decorators may be imported from other modules, decorators are specified to operate entirely runtime, rather than at early error time.
+
+## Runtime representation of decorators
+
+A decorator as applied is represented as a List of Decorator Records of the form { [[BuiltinDecorator]]: the String name of a built-in decorator, e.g., `"wrap"`, [[Arguments]]: a List of JavaScript values }.
```

Comment:
if `@` is part of the name, should it be part of this string as well?

Reply:
Yes, will fix (though this is editorial).